### PR TITLE
feat. sealos login hub.sealos.io by kubeconfig.

### PR DESCRIPTION
**Pull request by lingdie**
sealos login hub.sealos.uo by kubeconfig.

### CODE REVIEW

File: pkg/buildah/login.go [hunk 1]
```diff
@@ -75,7 +128,8 @@ func loginCmd(c *cobra.Command, args []string, iopts *loginReply) error {
 		return errors.New("too many arguments, login takes only 1 argument")
 	}
 	if len(args) == 0 {
-		return errors.New("please specify a registry to login to")
+		args = append(args, DefaultRegistry)
+		fmt.Println("login to default registry: hub.sealos.io.")
```
Comment (zzjin) [inline pkg/buildah/login.go:132]: Seems not necessary, as explained above.

File: pkg/buildah/login.go [hunk 2]
```diff
@@ -59,14 +68,58 @@ func newLoginCommand() *cobra.Command {
 		Use:   "login",
 		Short: "Login to a container registry",
 		Long:  loginDescription,
+
+		PreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			if len(args) == 0 {
+				args = append(args, DefaultRegistry)
+			}
+
+			if args[0] == DefaultRegistry {
```
Comment (zzjin) [inline pkg/buildah/login.go:77]: We can support any host wether they support `sealos` login or not.
Just parse user inputs, and do login, and save response.

If user want to login to `docker.io`, they can just use command: `sealos login docker.io -u xxx -p xxx` as usual, alone with user can login into `hub.sealos.io` witch currently support `sealos login hub.sealos.io -u xxx -k path/to/file`.

With this design, we can both support old behavior and plus support `kubeconfig` login. Further more, once our `hub.sealos.io` updated support  username+password login(which we can do but not do right now), users can login any registries as they need using the same method.

**Comment by cuisongliu** [comment]
I don't think it's appropriate to modify buildah's code directly. Is it necessary to add an applier layer for modification? This is the underlying logic. If you use this module in the startup place, these functions will be turned on by default.

File: pkg/buildah/login.go (hunk 2 above)
Comment (lingdie) [reply pkg/buildah/login.go:77]: > sealos login hub.sealos.io -u xxx -k path/to/file.
here, we can parse kubeconfig and get user.

**Comment by lingdie** [comment]
> I don't think it's appropriate to modify buildah's code directly. Is it necessary to add an applier layer for modification? This is the underlying logic. If you use this module in the startup place, these functions will be turned on by default.

We are not changing the functionality of the original function, we are just adding a flag to it. Originally there was pkg/image that did what you are talking about, but it was removed.

File: pkg/buildah/login.go [hunk 3]
```diff
@@ -48,6 +54,8 @@ func (opts *loginReply) RegisterFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&opts.tlsVerify, "tls-verify", opts.getLogin, "require HTTPS and verify certificates when accessing the registry. TLS verification cannot be used when talking to an insecure registry.")
 	fs.BoolVar(&opts.getLogin, "get-login", opts.tlsVerify, "return the current login user for the registry")
 	fs.AddFlagSet(auth.GetLoginFlags(&opts.loginOpts))
+	// e.g sealos login --kubeconfig /root/.kube/config hub.sealos.io
+	fs.StringVarP(&opts.kubeconfig, "kubeconfig", "k", opts.kubeconfig, "Login to sealos registry: hub.sealos.io by kubeconfig")
```
Comment (fanux) [inline pkg/buildah/login.go:58]: set a default value `~/.kube/config`

File: pkg/buildah/login.go [hunk 4]
```diff
@@ -59,14 +67,50 @@ func newLoginCommand() *cobra.Command {
 		Use:   "login",
 		Short: "Login to a container registry",
 		Long:  loginDescription,
+
+		PreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			// login by kubeconfig
+			if opts.kubeconfig != "" {
```
Comment (fanux) [inline pkg/buildah/login.go:73]: Return fast:
```
if opts.kubeconfig == "" {
    return ...
}
```

And if set the default value, opt.kubeconfig will not be ""

File: pkg/buildah/login.go [hunk 5]
```diff
@@ -59,14 +67,50 @@ func newLoginCommand() *cobra.Command {
...
+				// set user/password
+				var passwordb []byte
+				if passwordb, err = fileutil.ReadAll(opts.kubeconfig); err != nil {
```
Comment (fanux) [inline pkg/buildah/login.go:76]: The all kubeconfig content as password?

Comment (lingdie) [reply pkg/buildah/login.go:76]: yes.

File: pkg/buildah/login.go [hunk 6]
```diff
@@ -59,14 +67,51 @@ func newLoginCommand() *cobra.Command {
 		Use:   "login",
 		Short: "Login to a container registry",
 		Long:  loginDescription,
+
+		PreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			if opts.kubeconfig == "" {
+				return
+			} else {
```
Comment (zzjin) [inline pkg/buildah/login.go:74]: This if statement can simply delete `else`, for faster code return and better code Indent .

File: pkg/buildah/login.go [hunk 7]
```diff
@@ -90,3 +135,29 @@ func loginCmd(c *cobra.Command, args []string, iopts *loginReply) error {
...
+		return "", fmt.Errorf("failed to find current context %s", config.CurrentContext)
+	}
+	return expectedCtx.AuthInfo, nil
+}
+
+// If the specified string starts with http{s} it is replaced with it's
+// host[:port] parts; everything else is stripped. Otherwise, the string is
+// returned as is.
+func replaceURLByHostPort(repository string) (string, error) {
+	if !strings.HasPrefix(repository, "https://") && !strings.HasPrefix(repository, "http://") {
```
Comment (zzjin) [inline pkg/buildah/login.go:155]: Not as safe if passed by urls like 'ftp://sealos.io/../../root/.ssh/auth'.
 Maybe call `url.ParseRequestURI` and check return err.

File: pkg/buildah/login.go [hunk 8]
```diff
@@ -59,14 +67,51 @@ func newLoginCommand() *cobra.Command {
...
+				if err != nil {
+					return err
+				}
+				sealosKubeConfdir := fmt.Sprintf("%s/%s/%s/%s", os.Getenv("HOME"), ".sealos", registryHost, username)
+				err = fileutil.MkDirs(sealosKubeConfdir)
+				if err != nil {
+					return err
+				}
+				sealosKubeconfPath := fmt.Sprintf("%s/%s", sealosKubeConfdir, "config")
+				err = fileutil.Copy(opts.kubeconfig, sealosKubeconfPath)
```
Comment (zzjin) [inline pkg/buildah/login.go:99]: Seems we must wait login command succeeded, and then do the real copy action?

File: pkg/buildah/login.go (hunk 3 above)
Comment (zzjin) [reply pkg/buildah/login.go:58]: Seems not necessary. If we downloaded kubeconfig from our cloud site, the file will be in any location.

There is only one statement that`~/.kube/config` exists, witch is users using `sealos run` to start and manage a cluster, witch at this statement, they are not likely do login or push images, but  just pull and run sealos images.

File: pkg/buildah/login.go (hunk 6 above)
Comment (lingdie) [reply pkg/buildah/login.go:74]: ok...

File: pkg/buildah/login.go (hunk 8 above)
Comment (lingdie) [reply pkg/buildah/login.go:99]: I will move it to PostRunE.

File: pkg/buildah/login.go (hunk 7 above)
Comment (lingdie) [reply pkg/buildah/login.go:155]: Any legitimate url could be a registry, and it's hard to use the filesystem to distinguish between different registries or different sealos-cloud-instance.
For example:
- https://hub.sealos.io/registry1
- https://hub.sealos.io/registry2/part2
or without scheme
- hub.sealos.io:5000
or
- yy.hub.sealos.io.cn


I was wondering if we should also maintain a jason file like docker, e.g.
```jason
{
	"registrys": {
		"hub.sealos.io:35000": {
			"kubeconfig": "...",
                        "user": "...",
		},
	},
        "clouds":{
        }
}
```
so, we can just edit and read it to get kubeconfig.


Or Based on userid uniqueness just maintain file like:$(HOME)/.sealos/users/$(userid)/config
The probability of different sealos generating the same userid and using it on the same cluster is very low.

File: pkg/buildah/login.go [hunk 9]
```diff
@@ -59,14 +67,53 @@ func newLoginCommand() *cobra.Command {
 		Use:   "login",
 		Short: "Login to a container registry",
 		Long:  loginDescription,
+
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if opts.kubeconfig == "" {
+				return nil
```
Comment (fanux) [inline pkg/buildah/login.go:73]: need return not found error~

File: pkg/buildah/login.go (hunk 7 above)
Comment (zzjin) [reply pkg/buildah/login.go:157]: Use just `hostname` for path key is enough.

* http://hub.sealos.io/registry1 -> hub.sealos.io
* https://hub.sealos.io/registry2/part2 -> hub.sealos.io
* hub.sealos.io:5000 -> hub.sealos.io:5000
* yy.hub.sealos.io.cn -> yy.hub.sealos.io.cn

`http` and `https` refers to one domain.

We do not support one hotname with two or more registry, we can use:
* `registry1.hub.sealos.io`
* `registry2.hub.sealos.io` 

instead.

File: pkg/buildah/login.go [hunk 10]
```diff
@@ -59,14 +66,53 @@ func newLoginCommand() *cobra.Command {
...
+			registryHost, _, err := parseRawURL(args[0])
+			if err != nil {
+				return err
+			}
+			sealosKubeConfdir := fmt.Sprintf("%s/%s/%s/%s", os.Getenv("HOME"), ".sealos", registryHost, opts.loginOpts.Username)
+			if err := fileutil.MkDirs(sealosKubeConfdir); err != nil {
+				return err
+			}
+			sealosKubeconfPath := fmt.Sprintf("%s/%s", sealosKubeConfdir, "config")
+			if err := fileutil.Copy(opts.kubeconfig, sealosKubeconfPath); err != nil {
```
Comment (zzjin) [inline pkg/buildah/login.go:105]: Named `Copy` will override exists file, add comment to spec it.

**Review by zzjin — APPROVED** [review]

**Pull request merged**